### ISSUE TITLE
 Obsoleted by kernel fix

### ISSUE THREAD

**Issue by bdrung** [xkahn/camlink#6]
There is a bug in the firmware which breaks all pixel formats except the first one. I developed a quirk to fix the buggy firmware and submitted it to the upstream Linux: https://lore.kernel.org/lkml/20210325213458.51309-1-bdrung@posteo.de/

So once this patch is accepted, backported to older kernel released, and dribbled into the disto kernels this workaround library is not needed any more.

**Comment by xkahn** [comment]
@bdrung I'll keep this issue open for now, but I referenced your patch in the README

**Comment by jonhoo** [comment]
@bdrung Thank you for putting in the effort! Please do update here when the fix lands 🎉

**Comment by bdrung** [comment]
I will.

Regarding the README: The kernel patch adds a quirk to correct the broken response from the buggy firmware. With that patch, you can change the pixel format to any of those three available.

**Comment by bdrung** [comment]
Since I have three computers that need this kernel patch, I started maintaining this patched linux kernel in my PPA for Ubuntu 20.10 (groovy): https://launchpad.net/~bdrung/+archive/ubuntu/ppa

I plan to maintain this linux kernel in the PPA until the patch is included in the latest Ubuntu release.

**Comment by Snuggle** [comment]
@bdrung - Could I please ask if your patch has made it into the 21.04 release yet? I've looked through the mailing list but I'm not sure how to find which kernel version should include your patch. :blush:

Here is what formats are currently listed for me & current installed kernel:
```bash
snuggle@plum:~$ v4l2-ctl -d /dev/video0 --list-formats-ext
ioctl: VIDIOC_ENUM_FMT
	Type: Video Capture

	[0]: 'NV12' (Y/CbCr 4:2:0)
		Size: Discrete 3840x2160
			Interval: Discrete 0.040s (25.000 fps)
	[1]: 'NV12' (Y/CbCr 4:2:0)
		Size: Discrete 3840x2160
			Interval: Discrete 0.040s (25.000 fps)
	[2]: 'YU12' (Planar YUV 4:2:0)
		Size: Discrete 3840x2160
			Interval: Discrete 0.040s (25.000 fps)
snuggle@plum:~$ uname -a
Linux plum 5.11.0-17-generic #18-Ubuntu SMP Thu May 6 20:10:11 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by bdrung** [comment]
Sadly the patch is not included upstream yet and therefore also not submitted to Ubuntu. I sent the patch to the kernel mailing list four times, but got no response yet. You can follow the progress on https://lore.kernel.org/lkml/?q=Fix+pixel+format+change+for+Elgato+Cam+Link+4K

**Comment by bdrung** [comment]
Good news: The fix is now merged upstream and will be part of v5.14:
* fix commit: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=4c6e0976295add7f0ed94d276c04a3d6f1ea8f83
* merge: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=31e798fd6f0ff0acdc49c1a358b581730936a09a

The fix is also currently included in Ubuntu: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1932367

**Comment by bdrung** [comment]
So, the fix was backported to the stable releases. The fix is included in:
* 4.4.276
* 4.9.276
* 4.14.240
* 4.19.198
* 5.4.134
* 5.10.52
* 5.12.19
* 5.13.4
* 5.14-rc1

**Comment by GeorgeSapkin** [comment]
Has anybody tested the backported fixes?

**Comment by GeorgeSapkin** [comment]
OK, tested this with 5.13.4 on Fedora 34 with Firefox 90 and it seems to work without this repo's quirks. Good job! :tada:

**Comment by GeorgeSapkin** [comment]
I spoke too soon. Camera feed keeps freezing in Zoom with 5.13.4, not sure if this is related to this fix or it's something else in this kernel.

**Comment by bdrung** [comment]
I have been using this kernel fix for month without problems. If the camera feed is freezing during runtime and no pixel format change was triggered around this time, your issue has a different reason.

**Comment by GeorgeSapkin** [comment]
What reason might that be? Are you using Zoom as well for extended calls? Reverting to the previous kernel without this patch (5.12.15) seems to resolve it for me.

**Comment by bdrung** [comment]
Yes. I use Firefox and Chromium (jitsi meet and Google meet) and Zoom. Some calls run for multiple hours.

**Comment by GeorgeSapkin** [comment]
So, the camera feed freezes everywhere, including Firefox. What could it be?

**Comment by GeorgeSapkin** [comment]
The only related message that happens around the freeze is:
```
uvcvideo 5-1.2.1.1:1.1: Non-zero status (-71) in video completion handler.
```
